src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1281,13 +1283,100 @@
      (str "ALTER TABLE ONLY edges"
           "  ADD CONSTRAINT edges_certname_source_target_type_unique_key"
           "  UNIQUE (certname, source, target, type)"))))
+
 (defn add-latest-report-timestamp-to-certnames []
   (jdbc/do-commands
     "DROP INDEX IF EXISTS idx_certnames_latest_report_timestamp"
     "ALTER TABLE certnames DROP COLUMN IF EXISTS latest_report_timestamp"
     "ALTER TABLE certnames ADD COLUMN latest_report_timestamp timestamp with time zone"
     "CREATE INDEX idx_certnames_latest_report_timestamp ON certnames(latest_report_timestamp)"))
 
+(defn rededuplicate-facts []
+  (jdbc/do-commands
+   "CREATE SEQUENCE fact_values_id_seq;"
+
+   ;; will add not-null constraint on value_hash below; we don't have all the
+   ;; values for this yet
+   "CREATE TABLE fact_values (
+      id bigint DEFAULT nextval('fact_values_id_seq'::regclass) NOT NULL,
+      value_hash bytea,
+      value_type_id bigint NOT NULL,
+      value_integer bigint,
+      value_float double precision,
+      value_string text,
+      value_boolean boolean,
+      value jsonb not null
+    );"
+
+   "CREATE TABLE facts_transform (
+       factset_id bigint NOT NULL,
+       fact_path_id bigint NOT NULL,
+       fact_value_id bigint NOT NULL
+    );"
+
+   "INSERT INTO fact_values (value, value_integer, value_float, value_string, value_boolean, value_type_id)
+       SELECT distinct value, value_integer, value_float, value_string, value_boolean, value_type_id FROM facts"
+
+
+   "INSERT INTO facts_transform (factset_id, fact_path_id, fact_value_id)
+       SELECT f.factset_id, f.fact_path_id, fv.id
+         FROM facts f
+        INNER JOIN fact_values fv
+                ON fv.value_type_id = f.value_type_id
+               AND fv.value_integer IS NOT DISTINCT FROM f.value_integer
+               AND fv.value_float   IS NOT DISTINCT FROM f.value_float
+               AND fv.value_string  IS NOT DISTINCT FROM f.value_string
+               AND fv.value_boolean IS NOT DISTINCT FROM f.value_boolean
+               AND fv.value         = f.value")
+
+  ;; populate fact_values.value_hash
+  (jdbc/call-with-query-rows
+   ["select id, value::text from fact_values"]
+   (fn [rows]
+     (doseq [batch (partition-all 5000 rows)]
```

Comment:
on second thought, that's probably not useful for the batched update we need here